### ISSUE TITLE
S3Client can't use virtual hosted style for customize endpoint.

### ISSUE THREAD

**Issue by zmxnv1** [aws/aws-sdk-java-v2#2317]
## Describe the bug
I want to use S3 V2 SDK to visit aliyun oss. The aliyun oss only supports virtual hosted. 
But if I override the endpoint with "http://oss-cn-beijing.aliyuncs.com", the S3Client always chooses path style.

## Expected Behavior
S3Client chooses virtual hosted style to request server.

## Current Behavior
S3Client use path style.

## Steps to Reproduce
```java
S3Client client = S3Client.builder().endpointOverride(new URI("http://oss-cn-beijing.aliyuncs.com")).build();
client.putObject();   // the client will choose path style instead of virtual-hosted style.
```

## Possible Solution
![image](https://user-images.githubusercontent.com/18166125/110301150-b3e85500-8032-11eb-8c9e-e4d8011bb1a7.png)

It will choose virtual-hosted style only the host startswith "s3"

## Context
I can't use aws sdk to visit aliyun service.

## Your Environment

* AWS Java SDK version used: 2.15.67
* JDK version used: 11 Coretto 
* Operating System and version: macOS Big Sur.

**Comment by debora-ito** [comment]
@zmxnv1 that's the intended behavior, since the scope of the AWS SDK for Java is to work with AWS endpoints. We cannot guarantee that the SDK will work with third-party S3 applications. 

I'll change this to a feature request. PRs are welcome, although the solution would require some discussion.

**Comment by chunlinyao** [comment]
I migrated from aws sdk v1, this behavor is broken from v1.

```
    /**
     * Converts the current endpoint set for this client into virtual addressing style, by placing
     * the name of the specified bucket before the S3 service endpoint.
     *
     * @param bucketName The name of the bucket to use in the virtual addressing style of the returned URI.
     * @return A new URI, creating from the current service endpoint URI and the specified bucket.
     */
    private static URI convertToVirtualHostEndpoint(URI endpoint, String bucketName) {
        try {
            return new URI(String.format("%s://%s.%s", endpoint.getScheme(), bucketName, endpoint.getAuthority()));
        } catch (URISyntaxException e) {
            throw new IllegalArgumentException("Invalid bucket name: " + bucketName, e);
        }
    }
```
aws sdk v1 convert any valid dns endpoint url to virtual host style.

**Comment by yasminetalby** [comment]
Hello all,

This feature-request has been implemented as part of the 2.18.x version release.

As per the release Changelog, the new behavior is as follow: 

*When including an endpoint override, the SDK will use virtual bucket addressing as long as the bucket is a valid DNS label. This is different from the previous behavior where this only happens if the endpoint override begins with* `s3` : [see changelog](https://github.com/aws/aws-sdk-java-v2/blob/eb846d5689990c6f73ec105e7955f17c8cccf17f/CHANGELOG.md#L168)

As a side-note, if your bucket is not a valid DNS label , the behavior would be to not use virtual bucket addressing.

Thank you very much for your feature-request submission and feedback! 

Best,

Yasmine
